odl/space/cu_ntuples.py

```diff
@@ -807,6 +820,122 @@ def fun(x, out=None):
 sqrt = _make_unary_fun('sqrt')
 
 
+def _weighted(weight, attr, exponent, dist_using_inner=False):
+    if np.isscalar(weight):
+        weighting = CudaFnConstWeighting(
+            weight, exponent)
+    elif isinstance(weight, CudaFn.Vector):
+        weighting = CudaFnVectorWeighting(
+            weight, exponent=exponent, dist_using_inner=dist_using_inner,
+            copy_to_gpu=True)
+    else:
+        weight_ = np.asarray(weight)
+        if weight_.dtype == object:
+            raise ValueError('bad weight {}'.format(weight))
+        if weight_.ndim == 1:
+            weighting = CudaFnVectorWeighting(
+                weight_, exponent, dist_using_inner=dist_using_inner)
+        elif weight_.ndim == 2:
+            raise NotImplementedError('matrix weighting not implemented '
+                                      'for CUDA spaces.')
+#            weighting = CudaFnMatrixWeighting(
+#                weight_, exponent, dist_using_inner=dist_using_inner)
+        else:
+            raise ValueError('array-like weight must have 1 or 2 dimensions, '
+                             'but {} has {} dimensions.'
+                             ''.format(weight, weight_.ndim))
+    return getattr(weighting, attr)
```

Comment:
If that's hard to understand then look at [this class](https://github.com/odlgroup/odl-solvers/blob/issue-5__steplen/odlsolv/util/state.py) :grin:. I felt like doing some crazy shit on Friday...